player/javascript/defaults.js

```diff
@@ -0,0 +1,495 @@
+ - No id identity with symlinks - a valid choice which others make too.
+ - require("X") always maps to "X.js" -> require("foo.js") is file "foo.js.js".
+ - Global modules search paths are 'scripts/modules.js/' in mpv config dirs.
+ - A main script could e.g. require("./abc") to load a non-global module.
+ - Module id supports mpv path enhancements, e.g. ~/foo, ~~/bar, ~~desktop/baz
+ *********************************************************************/
+
+// Internal meta top-dirs. Users should not rely on these names.
+var MODULES_META = "~~modules",
+    SCRIPTDIR_META = "~~scriptdir",  // relative script path -> meta absolute id
+    main_script = mp.utils.split_path(mp.script_file);  // -> [ path, file ]
+
+function resolve_module_file(id) {
+    var sep = id.indexOf("/"),
+        base = id.substring(0, sep),
+        rest = id.substring(sep + 1) + ".js";
+
+    if (base == SCRIPTDIR_META)
+        return mp.utils.join_path(main_script[0], rest);
+
+    if (base == MODULES_META) {
+        var path = mp.find_config_file("scripts/modules.js/" + rest);
+        if (!path)
+            throw(Error("Cannot find module file '" + rest + "'"));
+        return path;
+    }
+
+    return id + ".js";
+}
+
+// Delimiter '/', remove redundancies, prefix with modules meta-root if needed.
+// E.g. c:\x -> c:/x, or ./x//y/../z -> ./x/z, or utils/x -> ~~modules/utils/x .
+function canonicalize(id) {
+    var path = id.replace(/\\/g,"/").split("/"),
+        t = path[0],
+        base = [];
+
+    // if not strictly relative then must be top-level. figure out base/rest
+    if (t != "." && t != "..") {
+        // global module if it's not fs-root/home/dos-drive/builtin/meta-dir
+        if (!(t == "" || t == "~" || t[1] == ":" ||  t == "@" || t.match(/^~~/)))
+            path.unshift(MODULES_META);  // add an explicit modules meta-root
+
+        if (id.match(/^\\\\/))  // simple UNC handling, preserve leading \\srv
+            path = ["\\\\" + path[2]].concat(path.slice(3));  // [ \\srv, shr..]
+
+        if (t[1] == ":" && t.length > 2) {  // path: [ "c:relative", "path" ]
+            path[0] = t.substring(2);
+            path.unshift(t[0] + ":.");  //  ->  [ "c:.", "relative", "path" ]
+        }
+        base = [path.shift()];
+    }
+
+    // path is now logically relative. base, if not empty, is its [meta] root.
+    // normalize the relative part - always id-based (spec Module Id, 1.3.6).
+    var cr = [];  // canonicalized relative
+    for (var i = 0; i < path.length; i++) {
+        if (path[i] == "." || path[i] == "")
+            continue;
+        if (path[i] == ".." && cr.length && cr[cr.length - 1] != "..") {
+            cr.pop();
+            continue;
+        }
+        cr.push(path[i]);
+    }
+
+    if (!base.length && cr[0] != "..")
+        base = ["."];  // relative and not ../<stuff> so must start with ./
+    return base.concat(cr).join("/");
+}
```

Comment:
I have no idea what it does, but it looks pretty complicated, and Lua doesn't seem to need it.